ark/spLDA/processing.py

```diff
@@ -63,24 +52,153 @@ def format_cell_table(cell_table, markers=None, clusters=None, fovs="all_fovs"):
-        df['isimmune'] = True
+        df["is_index"] = True
+        df["is_immune"] = True  # might remove this
         fov_dict[i] = df.reset_index(drop=True)
 
-    return(fov_dict)
+    # Save Arguments
+    fov_dict["fovs"] = fovs
+    fov_dict["markers"] = markers
+    fov_dict["clusters"] = clusters
+
+    return fov_dict
 
 
+def featurize_cell_table(cell_table, feature_by="cluster", radius=100,
+                         cell_index="is_index", n_processes=None,
+                         train_split=True):
+    """Calculates statistics for local cellular neighborhoods based on the
+    specified features and radius.
 
+    Args:
+        cell_table (dict, pd.DataFrame):
+            A formatted cell table for use in spatial-LDA analysis.
+            Specifically, this is the output from
+            format_cell_table().
+        feature_by (str):
+            One of four choices of featurization method:
+                * marker: for each marker, count the total number of cells
+                within a ``radius`` *r* from cell *i*
+                having marker expression greater than 0.5.
+                * avg_marker: for each marker, compute the average marker
+                expression of all cells with a ``radius`` *r*
+                from cell *i*.
+                * cluster: for each cluster, count the total number of cells
+                with a ``radius`` *r* from cell *i*
+                belonging to that cell cluster.
+                * count: counts the total number of cells within a ``radius``
+                *r* from cell *i*.
+        radius (int):
+            Size of the radius, in pixels, used to featurize cellular
+            neighborhoods.
+        cell_index (str):
+            Name of the column containing the reference cell indexes.  If not
+            specified, all cells are used.
+        n_processes (int):
+            Number of parallel processes to use.
+        train_split (bool):
+            Create an index for training data in each field of view.
 
+    Returns:
+        dict:
 
+        - A dictionary containing featurized cellular neighborhoods and
+        training splits if applicable.
+    """
+    check_featurize_cell_table_args(cell_table=cell_table,
+                                    feature_by=feature_by, radius=radius,
+                                    cell_index=cell_index)
+    # Define Featurization Function
+    func_type = {
+        "marker": neighborhood_to_marker,
+        "cluster": neighborhood_to_cluster,
+        "avg_marker": neighborhood_to_avg_marker,
+        "count": neighborhood_to_count
+    }
+    if feature_by in ["marker", "avg_marker"]:
+        neighborhood_feature_fn = functools.partial(func_type[feature_by],
+                                                    markers=cell_table[
+                                                        "markers"])
+    else:
+        neighborhood_feature_fn = functools.partial(func_type[feature_by])
```

Comment:
I understand we're pulling from a dict of functions, but what is this snippet doing?

Reply:
I pulled this from the Calico Labs example notebook.  I believe it's necessary for certain steps in the featurize_samples() function.

Reply:
It looks like the "marker" and "avg_marker" functions require the `markers` argument, whereas "cluster" and "count" do not

Reply:
that's also true